src/gluonts/dataset/loader.py

```diff
@@ -57,52 +62,51 @@ def __init__(
         batch_size: int,
         ctx: mx.Context,
         dtype: DType = np.float32,
+        num_workers: int = None,
+        cyclic: bool = False,
+        **kwargs
     ) -> None:
+        # conversion from iterator to list
         self.batch_size = batch_size
         self.ctx = ctx
         self.dtype = dtype
         self.is_train = is_train
-        self.dataset = dataset
         self.transform = transform
-
-    @property
-    def stream(self) -> Iterable:
-        return self.transform(self.dataset, is_train=self.is_train)
-
-    def make_batch_iter(self):
-        batches = batcher(self.stream, self.batch_size)
-        stack = functools.partial(dct_reduce, self.stack)
-        return map(stack, batches)
-
-    def stack(self, xs):
-        if isinstance(xs[0], np.ndarray):
-            data = np.asarray(xs)
-            if data.dtype.kind == "f":
-                data = data.astype(self.dtype)
-            return mx.nd.array(data, dtype=data.dtype, ctx=self.ctx)
-
-        if isinstance(xs[0], mx.nd.NDArray):
-            return mx.nd.stack(*xs)
-
-        if isinstance(xs[0], list):
-            return list(self.stack(t) for t in zip(*xs))
-
-        if isinstance(xs[0], tuple):
-            return tuple(self.stack(t) for t in zip(*xs))
-
-        return xs
+        self.cyclic = cyclic
+
+        # TODO: think about what a good value is, probably 0, and if multiprocessing=True, then what is below
+        if num_workers is None:
+            self.num_workers = min(len(list(dataset)), int(cpu_count() / 2))
+        else:
+            assert num_workers <= len(
+                list(dataset)
```

Comment:
Why `list` here? Don't we have special implementation for `len` on `Dataset` that are much faster?